File: src/snippets/guards/quick-start/langchain-js/Wrap.ts
import { launchArcjet, localDetectSensitiveInfo } from "@arcjet/guard";
import { rampart } from "@arcjet/sensitive-info-rampart";
import { guardMiddleware, guardTool } from "@arcjet/guard/langchain/v1";
import { createAgent } from "langchain";
import { tool } from "@langchain/core/tools";
import { z } from "zod";

// Create one Arcjet client and reuse it across agent runs. Rampart
// detects bank account and routing numbers locally.
const arcjet = launchArcjet({
  key: process.env.ARCJET_KEY!,
  sensitiveInfoBackend: rampart(),
});
const detectPii = localDetectSensitiveInfo({
  deny: ["BANK_ACCOUNT", "ROUTING_NUMBER"],
});

export async function runEmailAgent(
  user: {
    id: string;
    allowedRecipients: string[];
    record: {
      name: string;
      bankAccount: string;
      routingNumber: string;
    };
  },
  prompt: string,
) {
  const getClientRecord = tool(async () => user.record, {
    name: "get_client_record",
    description:
      "Get the account details on file for the current customer",
    schema: z.object({}),
  });

  // This adapter accepts action and rules. It doesn't accept
  // inputs.
  const sendEmail = guardTool(
    arcjet,
    tool(
      async ({ recipient, body }) =>
        emailProvider.send({ to: recipient, body }),
      {
        name: "send_email",
        description: "Send an email",
        schema: z.object({
          recipient: z.string().email(),
          body: z.string(),
        }),
      },
    ),
    {
      action: "email.sent",
      rules: ({ body }) => [detectPii(body)],
    },
  );

  const agent = createAgent({
    model: "openai:gpt-4o-mini",
    tools: [getClientRecord, sendEmail],
    middleware: [guardMiddleware(arcjet, { sessionId: user.id })],
  });

  return agent.invoke(
    { messages: [{ role: "user", content: prompt }] },
    { configurable: { thread_id: user.id } },
  );
}

File: src/snippets/guards/quick-start/langchain/Wrap.py
import os

from arcjet.guard import launch_arcjet, local_input, server_input
from arcjet.guard.langchain import guard_tool
from arcjet_sensitive_info_rampart import rampart
from langchain.agents import create_agent
from langchain_core.tools import tool
from langchain_openai import ChatOpenAI

# Create one Arcjet client and enable local bank-detail detection.
arcjet = launch_arcjet(
    key=os.environ["ARCJET_KEY"],
    sensitive_info_backend=rampart(),
)


@tool
async def send_email(recipient: str, body: str) -> str:
    """Send an email."""
    await email_provider.send(to=recipient, body=body)
    return "sent"


async def run_email_agent(user, prompt: str):
    # This read-only tool returns the current customer's account
    # data.
    @tool
    async def get_client_record() -> dict:
        """Get the account details on file for the current customer."""
        return user.record

    # guard_tool checks remote policy before send_email can run.
    guarded_send_email = guard_tool(
        guard=arcjet,
        tool=send_email,
        # The action selects the remote policy configured in step 1.
        action="email.sent",
        # Actor and the allow list come from trusted application
        # state.
        actor=user.id,
        # Map only the values the remote policy needs.
        inputs=lambda arguments, _config: {
            "recipient": server_input.string(arguments["recipient"]),
            "allowed_recipients": server_input.string_list(
                user.allowed_recipients
            ),
            "body": local_input.string(arguments["body"]),
        },
    )

    # The model can call either tool, but send_email always passes
    # through Arcjet.
    agent = create_agent(
        ChatOpenAI(model="gpt-4o-mini"),
        tools=[get_client_record, guarded_send_email],
        system_prompt=(
            "Use get_client_record when the user asks for account "
            "details. Use send_email exactly once to complete the "
            "request."
        ),
    )
    return await agent.ainvoke(
        {"messages": [{"role": "user", "content": prompt}]}
    )
